float64 range
int32 angle